### ISSUE TITLE
Sharing 2.0

### ISSUE THREAD

**Issue by rullzer** [owncloud/core#19331]
@schiesbn and myself spend some time at the conference coming up with a possible design for sharing 2.0. Here is basically an overview of my notes.

## Note

As discussed in Berlin the sharing stuff will for now only be for files. Contact and calendar will use the old sharing code but will be moved to webdav.

## Architecture

We want to move all real sharing to separate apps. Right now sharing is mostly in core and this is not really the right place. It makes it hard to have custom sharing implementations.

We propose a single `sharemanager` in core where apps can register themself. Furthermore in core we will provide a number of interface that apps have to implement in order to become a `shareprovider`. 

Currently we have the following 3 categories of shares:
* user/group
* link
* federated

We will of course keep those 3 categories but each of them will be a separate app. This would mean we would get more core apps. Which will take some time but in the end we should end up with a more modular (and better maintainable) sharing setup.

We will then also need another new app that handles the OCS Sharing API. This clearly separates the API and the sharing implementation 

### No more sharing hierarchy

We propose to move to what we like to call a flat sharing model. This will prevent cyclic sharing or at least make sure it is no longer an issue. (see also https://github.com/owncloud/core/issues/9058)

This effectively means that shares are a combination of the following items
* path
* share owner (owner of the path)
* sharer (initiatior of the share)
* sharee (receiver of the share)
* permissions

This means that the share owner always has the complete overview of who has access to a share.
And that we can easily find out the max permissions for a sharee.

#### sharemanager

The share manager is just a place where `shareproviders` can register and where they can be queried. There will be some functions to get a specific share or to get a traversal of all shares in a path etc.

#### shareprovider

The share provider just implements sharing for a specific type. You can of course tell the provider to share a file. Or to get a share etc.

Each share provider has a unique id.

We plan to have two share providers:

1. "internal share provider" which will be provided by the primary storage and handle (user, group and link sharing)
2. "federated share provider".

In general share provider will register them self at the share manager. This way it is easy to develop and add new share provider. For the internal share provider (which could be seen as the main share provider) the share manager will ask the primary storage directly to provide it.

We will implement two share providers as apps, the "oc internals share provider" which does sharing basically as we are doing it today and the "oc federated share provider". Bot will register them self at the share manager if the apps are activated. Additionally the share manager will ask the primary storage if he implements his own share provider. If he returns a share provider it will be used for internal shares, if it returns false the "oc default internal share provider" will be used. That's also the reason why the "oc default share provider" needs to be implemented as a app and not as part of our default storage, because the share provider needs to be also available for other storage if they don't implement their own share provider.

#### share

Just an object representation of the share. Where you can get and set the attributes etc.

Each share withing a shareprovider has a unique id.

A system wide unique id could be constructed as "providerID#localShareID" or something similiar.

We decided to only pass around userids/groupids/federatedshare ids etc. This makes sure that only the places that have to care about existing users etc will handle that.

#### Example Code

I took the liberty of already starting with some interfaces:

[ShareManager](https://github.com/rullzer/core/blob/sharing2.0_start/lib/public/share/imanager.php)
[ShareProvider](https://github.com/rullzer/core/blob/sharing2.0_start/lib/public/share/iprovider.php)
[Share](https://github.com/rullzer/core/blob/sharing2.0_start/lib/public/share/ishare.php)

## Limitation

We can only handle 1 share provider per type.

During the conference we had the idea to do things on a per storage level. But this brings in a lot of problems. Assume there are 2 folders `dirA` and `dirB`. Now the storage where `dirA` lives supports link sharing but the storage of `dirB` does not. Now what happens if a file with link share is moved from `dirA` to `dirB`?

In general most people will just use the apps that we provide. And in the case that people want very specific sharing behaviour for certain storages they should write their own sharing apps. 

Forcing this limitations keeps things simple and understandable and we do not believe this to be a limitation a lot of people will have problems with.

## Migration/deprecation path

For third party apps like calendar and contact we will move the existing sharing code to the apps and will make them work. From this point on the app delevlopers have to maintain or replace the existing code.

For file sharing we will re-use the exiting table, update the values add needed columns and remove old columns we no longer need at the end. The idea is that this process will be triggered with a occ script after the main upgrade to 9.0 while the system is live (still need to evaluate if this is possible).

Open Question:

- [ ] What happens for small installations with no access to a terminal? We probably need a fall back to trigger this update in the admin settings (similar to what we do for encryption key migration)
- [ ] When we migrate the shares step by step on a running system re-shares will not be visible until they are migrated (sync clients will potentially delete local copys and download them later again if the re-share shows up again)

## Improvements

Since we kind of start over. There are a lot of things where we can improve and unify the current sharing design. Some things we came up with.

### Expiration date on everything

Currently it is only possible to have expiration dates on link shares. There is no real technical reason for this and it is perfectly valid to have other shares expire as well. 

### Multiple link shares

Currently only 1 link share is allowed per file. Having multiple makes sense since I might want to share a file with multiple people but not for the same period of time. This could be combined with easily sharing via email so the individual link is automatically sent to the respective person. Individual links can then also be revoked easily based on email/name.

## Open Issues

- [ ] decide about the right default permission for shares https://github.com/owncloud/core/issues/11707#issuecomment-143165474

## Checklists

- Share backend
  * [ ] Share manager ( https://github.com/owncloud/core/pull/19872)
    * [x] Throw delete hooks (https://github.com/owncloud/core/issues/20543)
    * [x] Delete all children of share on delete (https://github.com/owncloud/core/issues/20543)
    * [ ] allow to pass in a token when creating a new share
    * [x] Allow registration of providers (https://github.com/owncloud/core/pull/21552)
    * [x] Move interfaces to OCP (https://github.com/owncloud/core/pull/21957)
  * [x] Default share Provider
    * [x] DB modifications (https://github.com/owncloud/core/pull/20703)
    * [x] List share (https://github.com/owncloud/core/pull/19872, https://github.com/owncloud/core/pull/20356)
    * [x] Delete share (https://github.com/owncloud/core/pull/19872)
    * [x] Get children (https://github.com/owncloud/core/pull/19872)
    * [x] Update Share (https://github.com/owncloud/core/pull/21887)
    * [x] Create Share (https://github.com/owncloud/core/pull/20773)
  * [x] Federated Share Provider (https://github.com/owncloud/core/pull/22080)
    * [x] List share 
    * [x] Delete share
    * [x] Get children
    * [x] Update Share
    * [x] Create Share
- Move OCS Share API to new sharing code
  * [x] Delete Shares (https://github.com/owncloud/core/pull/19872)
  * [x] List single shares (https://github.com/owncloud/core/pull/20356)
  * [x] List multiple shares (https://github.com/owncloud/core/pull/21532)
  * [x] List shares shared with me (https://github.com/owncloud/core/pull/21532)
  * [x] Update shares
     * [x] User shares (https://github.com/owncloud/core/pull/21887)
     * [x] Group shares (https://github.com/owncloud/core/pull/21887)
     * [x] Link Shares (https://github.com/owncloud/core/pull/21887)
     * [x] Federated Shares (https://github.com/owncloud/core/pull/22080)
  * [X] Create Shares (https://github.com/owncloud/core/pull/20773)
    * [X] User Share (https://github.com/owncloud/core/pull/20773)
    * [X] Group Share (https://github.com/owncloud/core/pull/20773)
    * [X] Link Share (https://github.com/owncloud/core/pull/20773)
    * [x] Federated share (https://github.com/owncloud/core/pull/22080)
- Other code
 * [ ] Move internal sharing calls to the share manager
    * [X] Link shares (https://github.com/owncloud/core/pull/21761)
    * [ ] Sharedmount etc
    * [ ] Federated sharing accepting etc
 * [x] Move webinterface over to the OCS Share API (#21554)
 * [x] Migration of old shares (https://github.com/owncloud/core/pull/21660)
 * [x] Provide more public exceptions (https://github.com/owncloud/core/pull/21957#issuecomment-176081619)

## How we implement it

The idea is to implement it in parallel to the existing sharing with small pull request merging sharing 2.0 one by one. In this transition phase it will be possible to switch between old and new implementation for testing purpose. Only if sharing 2.0 is complete and works reasonable well (including all integration tests for the OCS API) we will move the old sharing to the calendar and contacts app like explained under "Migration" and switch permanently to the new sharing.

## DB implentation and considerations

In order to faciliate the new sharing we need to modify the database. We want to do this as little as possible to avoid large and huge migration steps.

Eventually we will remove the columns that are only needed for calendar/contacts sharing since they move to webdav. But this can only be done once we kill off the old code (since it is heavily wired into it).

Currently we store the initiated of the share in `uid_owner`. So the table `uid_owner` maps to the sharer (sharedBy). We can get the owner of a file via the fileid in the table. But this does not scale. Thus we plan to introduce a new column that holds the owner of the file (shareOwner). This is then allows for quick lookups of share info for the owner since the owner of a file should always have the full overview.

## Related issues:
- https://github.com/owncloud/core/issues/13014 : previous sharing 2.0 issue
- https://github.com/owncloud/core/issues/17861 : Allow filtering of shares list
- https://github.com/owncloud/core/issues/13202 : Share permission issues
- https://github.com/owncloud/core/issues/9058 : Resharing behaviour
- https://github.com/owncloud/core/issues/5491 : Avoid loop shares
- https://github.com/owncloud/core/issues/20582 : OCS API /shares/id should also return for shares with me

**Comment by rullzer** [comment]
FYI: @DeepDiver1975 @karlitschek @cmonteroluque @MTRichards

**Comment by DeepDiver1975** [comment]
> We propose a single sharemanager in core where apps can register themself. Furthermore in core we will provide a number of interface that apps have to implement in order to become a shareprovider.

I question the use case - what would such a share provider be? 
If sharing is file only - this is a pure private implementation - no app needs to interact on this level.

**Comment by schiesbn** [comment]
> I question the use case - what would such a share provider be? 

A share provider could be a sharing implementation from a specific storage for example.
We need a manager to track the share provider (we may have more than one in the future) and redirect for example OCS share api calls to the right implementation.

**Comment by DeepDiver1975** [comment]
> A share provider could be a sharing implementation from a specific storage for example.

Which we need to get from the storage interface from my pov

**Comment by schiesbn** [comment]
> Which we need to get from the storage interface from my pov

We will see if we need it or not if we move forward with the implementation. But if we assume that not every storage will have is own implementation of sharing we probably need some kind of manager as entry point. For example you have a OCS call change permission to read-only for share "42", then we need a central instance (the share manager) who says OK share "42" is managed by share provider X (which might come from storage Y) and forward the request to change the permissions. How should a OCS share request for share "42" find the right storage by itself?

**Comment by rullzer** [comment]
> Which we need to get from the storage interface from my pov

I think we should have a clear separation. If you have multiple storage providers there is no guarantee that they support the same set of sharing functionality.

**Comment by DeepDiver1975** [comment]
hmmm .. true ... I guess we need both mechanisms.

On the one hand there is the need to explicitly share a file/folder on storage level

```
php
$storage->getSharing()->share()
```

on the other hand there has to be a mechanism to resolve a share to the real physical resource

```
php
$shareManager->resolveShare('42')
```

**Comment by DeepDiver1975** [comment]
> I think we should have a clear separation. If you have multiple storage providers there is no guarantee that they support the same set of sharing functionality.

There will be no new kinds of sharing.

And even if this is to be analysed how this might fit into the existing structures

**Comment by rullzer** [comment]
> > We propose a single sharemanager in core where apps can register themself. Furthermore in core we will provide a number of interface that apps have to implement in order to become a shareprovider.

> I question the use case - what would such a share provider be?
> If sharing is file only - this is a pure private implementation - no app needs to interact on this level.

A share provider for 

* user/group sharing 
* link sharing
* federated sharing

And maybe more in the future. This way if for example somebody wants to implement their own user/groups sharing on a file metadata level. But still wants to have link sharing. They only need to write their own user/group sharing app.

**Comment by schiesbn** [comment]
@DeepDiver1975  Yes, we need both. I think this are two different levels of abstraction. The share manager is on a higher level and manage sharing no matter which kind of storage or share implementation is used. It will then resolve the storage/share implementation which has to handle the share and call the share implementation of the storage to perform the action.

**Comment by DeepDiver1975** [comment]
@schiesbn sounds reasonable

But never the less I still question the share provider use case

**Comment by PVince81** [comment]
What happened to the similarly named ticket #13014 ?

**Comment by schiesbn** [comment]
#13014 was a early brain storming. This one is far more concrete and will be the issue where we define sharing for 9.0 and track the progress. I will check the other issue and see what we can/should move over here.

**Comment by karlitschek** [comment]
Looks good to me.
We have to keep the upgrade time under control as discussed in the zero upgrade discussion.
So we either make the new code also work with the old sharing tables and run the migration script in the background? Or we make sure that the migration script runs in minutes even for huge installations.

**Comment by DeepDiver1975** [comment]
> So we either make the new code also work with the old sharing tables

Are there any reasons (for now) to introduce a new table?

**Comment by karlitschek** [comment]
@DeepDiver1975 Excellent point. I assumed that there will be changes in the data model. Everything would be a lot easier if this can be avoided of course.

**Comment by rullzer** [comment]
> Are there any reasons (for now) to introduce a new table?

I don't know yet. We have to see when we get a bit more into detail. I think yes. since we want to move from a hierarchical sharing model to a flat sharing model (so the owner of the data always knows who has access etc). Doing this live in the same table as the current one is asking for complicated code.

But as I said I'm simply do not know yet. We probably should have a more in depth meeting at some point and maybe even have some (simple) code ready to look at so the discussion is not purely on a high level.

**Comment by jancborchardt** [comment]
Also keep in mind sharing via email. Normally this is just sending the share link to a person. You should be able to simply type in a person’s email (autocompletion from Contacts as well) and then ownCloud will generate an individual link for that person and also show them in the share list.

That way you can also remove these people from the share. In the backend it would just invalidate the custom link of that person – for others it would still work.

**Comment by schiesbn** [comment]
cc @cmonteroluque as discussed I updated the issue with the latest state (all important information should be in the initial comment https://github.com/owncloud/core/issues/19331#issue-108085914)

@rullzer @DeepDiver1975 please also review it. I updated the sections "share provider", "migration" and added the section "how we implement it"

**Comment by cmonteroluque** [comment]
thanks @schiesbn!

**Comment by schiesbn** [comment]
Updated the "shareprovider" section at the top to discuss the need of a way to register share providers at the manager and the rationals why both the "oc default internal share provider" and the "federated share provider" will be a app outside of the storage.

**Comment by nickvergessen** [comment]
Just a comment, we need a php internal API for shares. I just spoke to @schiesbn and he said that this was not the plan anymore. Quick example where we need it: Activities

When writing a file, we surely do not want to make a OCS request to the share API to get all users that have this file shared.

**Comment by DeepDiver1975** [comment]
> When writing a file, we surely do not want to make a OCS request to the share API to get all users that have this file shared.

can you please link the current code which is taking care of this?

**Comment by nickvergessen** [comment]
Activity app listens to the following [hooks](https://github.com/owncloud/activity/blob/master/lib/fileshooksstatic.php#L36-L40):
```php
		Util::connectHook('OC_Filesystem', 'post_create', 'OCA\Activity\FilesHooksStatic', 'fileCreate');
		Util::connectHook('OC_Filesystem', 'post_update', 'OCA\Activity\FilesHooksStatic', 'fileUpdate');
		Util::connectHook('OC_Filesystem', 'delete', 'OCA\Activity\FilesHooksStatic', 'fileDelete');
		Util::connectHook('\OCA\Files_Trashbin\Trashbin', 'post_restore', 'OCA\Activity\FilesHooksStatic', 'fileRestore');
		Util::connectHook('OCP\Share', 'post_shared', 'OCA\Activity\FilesHooksStatic', 'share');
```
Then I grab the [original file owner for the file](https://github.com/owncloud/activity/blob/master/lib/fileshooks.php#L185-L200):
```php
		$uidOwner = Filesystem::getOwner($path);
		$fileId = 0;
		if ($uidOwner !== $this->currentUser) {
			Filesystem::initMountPoints($uidOwner);
		}
		$info = Filesystem::getFileInfo($path);
		if ($info !== false) {
			$ownerView = new View('/' . $uidOwner . '/files');
			$fileId = (int) $info['fileid'];
			$path = $ownerView->getPath($fileId);
		}
```

And finally I get all users that share the file:
https://github.com/owncloud/activity/blob/master/lib/fileshooks.php#L176

**Comment by DeepDiver1975** [comment]
okay - this is something we need to consider - THX

@rullzer @schiesbn this still fits into our agreed approach to define the ocp interface in a later stage

**Comment by rullzer** [comment]
Made a start with a bunch of checklist of stuff that we need to do. This list might expand more but it helps us keep track.

@MTRichards to keep you updated. The plan now is to have small incremental improvements to (as a first step) improve the OCS Share API . Since that would also be what the web UI will start using. During this we work with QA to make sure we don't break API and stuff keeps working. So hopefully we will have a steady stream of small PR's.

**Comment by MTRichards** [comment]
Excellent. :) Thanks for the list!

**Comment by MTRichards** [comment]
As a question, does ShareManager allow for reporting, or is that out of scope?
We get requests for creating share reports based on a user, folder or file to know who shared what with whom. If we had a core API we could query, we could just use that and build an app.
Alternatively, we just query the database, but that is more likely to break and we should abstract that if we can.

Note: This is NOT a requirement, just a question.

**Comment by rullzer** [comment]
THose kind of queries should be possible. But maybe we should have a chat soonish what you actually want (some use cases) so we can see if we can incorporate that in the future.

**Comment by DeepDiver1975** [comment]
> As a question, does ShareManager allow for reporting, or is that out of scope?
> We get requests for creating share reports based on a user, folder or file to know who shared what with whom. If we had a core API we could query, we could just use that and build an app.
> Alternatively, we just query the database, but that is more likely to break and we should abstract that if we can.
> 
> Note: This is NOT a requirement, just a question.

@MTRichards generally speaking: anything is possible - can we postpone this so that we can for now focus on getting 9.0 finished? THX

**Comment by MTRichards** [comment]
Certainly, not a problem. I just want to plant it in your head - sometimes thinking ahead of time prevents decisions that create problems later.

Like I said:
>Note: This is NOT a requirement, just a question.

**Comment by DeepDiver1975** [comment]
Im happy to chat about this anytime.

**Comment by PVince81** [comment]
@MTRichards performing inception ? :wink: 

Ideal would be to raise a separate ticket to start a discussion about this possible enhancement.

**Comment by MTRichards** [comment]
https://github.com/owncloud/core/issues/20492
Done!

**Comment by rullzer** [comment]
Intergration tests are tracked in https://github.com/owncloud/core/issues/20553

**Comment by rullzer** [comment]
@DeepDiver1975 @schiesbn what do we want to do regarding ids? Currently id's are ints. However if we have multiple providers (which we will) then just ints is not enough. Since that can lead to collisions.

Now there are 2 approaches I see:

1. Do <shareid><sharetype>. So assume I have a user share with id 42. Then this would get id 4200 for example. And a federated id with id 13 would become 1306.

This is very simple and will make sure that the id stays an int. But it will force shareproviders to use ints as id.

Of course we could just use strings. So

2. <providerid>:<shareid>. Assume a federated share with id 8 would then become e.g. `fed:8`, a group share with our default provider with id 10 `ocdef:10`. Or a link share with some custom implentation might be `custom:c8fff9a`.

The second one is more flexible but might require changes to consumers of the OCS Share API (e.g. mobile/desktop clients).

I think in the long run the second option might be the most flexible. But I'm not sure. Opinions/suggestions?

**Comment by DeepDiver1975** [comment]
both approaches above will result in a string datatype because 
- there has to be a separator in both cases from my pov
- the simplistic int int concat will potentially break the int range

But most importantly: do we consider the change of an share id as public api breakage?

example:
before 9.0 the share id was 12345
after 9.0 the share is will be "default:12345"

summoning @karlitschek for api changes
summoning @dragotin @rperezb for the clients

**Comment by rullzer** [comment]
Well of course the providers don't have to store their own ID in the databae. Than can just be added when the object is created. So we can (and should!) still use ints for our internal numbering. Since generating unique id's is very non trivial.

If it is breakage kind of depends on your definition. We do say on https://doc.owncloud.org/server/8.2/developer_manual/core/ocs-share-api.html that the shareid is an int. However if we tell the clients in advance they can just use a string for that field and this should not break current behaviour since the int can just be stored as a string.

It is even a string when passed to the API. We just convert it back to an int.
The id is internal anyway and should never be shown to the user.

**Comment by rullzer** [comment]
@dragotin  FYI, while writing the new sharing code for the client I already made the id a string (https://github.com/owncloud/client/blob/master/src/gui/share.h#L111). So nothing to do for the desktop client.

**Comment by karlitschek** [comment]
@DeepDiver1975 good question. int vs. string makes definitely a different in client depending on the technology. If this would be a 'write only' variable then it would be ok because it is still backwards compatible. Unfortunately we also return the share id in several API calls. So this is a problem. :-(
Let's talk about this in the next call.

**Comment by rullzer** [comment]
@DeepDiver1975 suggested to use a wrapper in the OCS endpoint. So if you only pass an int it is by default an OC internal share. And gets mapped to `default:int`. Maybe something to consider.

**Comment by rullzer** [comment]
@DeepDiver1975 trouble is that fails when fetching shares. because there we'll return the new id. Since we have no way of knowing if the client support the new ids this will sometimes fail.

**Comment by DeepDiver1975** [comment]
well - v1 of the sharing api is only for non-federated shares and as a result no provider prefix is required - right?

@schiesbn

**Comment by rullzer** [comment]
Not entirely true. Currently we do return outgoing federated shares in the oc_share table (and thus return them via the OCS Share API).

Also this would then mean that all share providers must use ints as identifiers.

**Comment by schiesbn** [comment]
Maybe we could keep the ids as ints if we make sure that the request always comes with two information, the id and the share type instead of combining both information in the id. But this would be probably also a API change because not all api calls contain both information separately, id and share type.

Further I'm not sure if it work out well if we limit the id to ints, maybe some storages will generate ids in a different way. With strings we would be more flexible.

**Comment by DeepDiver1975** [comment]
> Further I'm not sure if it work out well if we limit the id to ints, maybe some storages will generate ids in a different way. With strings we would be more flexible.

from my pov we will need string because of all discussed reasons - but this would require v2 of the sharing api which I'd like to postpone for now (maybe a topic for 9.1).

We should not add too much changes at a time

**Comment by rullzer** [comment]
New issue that popped up.

1. user0 shares a folder with user1
2. user1 reshares a subfolder of that with user2

So far so good. So user0 will see

```
folder (shared with user1)
|-subfolder (shared with user2 by user1)
```

3. user0 removes the share of folder with user1

Now what happens? Is the share of subfolder with user2 automatically removed. The thing here is that it can be very costly to traverse the whole tree below a share to find subshares.

This is somewhat inconsistent with the flat sharing model. Since if user1 would just share folder with user2. It would be flattend out. And removing the share with user1 does not harm.

Some suggestions that came up:

* Leave the reshare. It will showup in API calls but that was the state then.
* Traverse the `folder`s children and remove all shares of subfolders.
* Traverse the `folder`s children and update all the shares of subfolder.
* Remove the reshares of subfolders possibility

Removing of the reshare functionality of subfolder is by far the simplest solution. Which will remove a lot of complexity. But also limits the user experince. If somebody shares a photo album with me and I want to share one of those pictures with a link. This would then be no longer possible.

Leaving the reshare as is feels weird. Since well user1 no longer has access to the files.

Traversing could be an option but might be potentially costly. ( @icewind1991 can you comment on this?)

Comments? @DeepDiver1975 @schiesbn

**Comment by icewind1991** [comment]
I think removing the reshare is correct.

As for performance maybe keep track of the relation between a share and all reshares made from that in the db

**Comment by MTRichards** [comment]
Yep, remove the reshare because you don't have access to the folder anymore.

**Comment by schiesbn** [comment]
I think removing the re-shares of a sub-file or sub-folder is OK.
But what happens if the top-folder was re-shared?

Following situation:

- user1 shares folder1 with user2
- user2 re-shares the same folder1 with user3
- if user1 looks at the share dialog of folder1 he will see that the folder is shared with user2 and user3
- now he unshares the folder from user2

If we remove the re-shares as well user3 will also disappear from the list of recipients, although user1 unshared the folder only from user2. Wouldn't this be a strange behavior?

With the  new way of handling shares, a flat list instead of a tree of re-shares I think it would be fine if we don't try to unshare re-shares. The owner sees all shares, so he is in full control. No need to "magically" remove additional shares just because a parent folder was unshared. Sure this gives us the challenge described by @rullzer, how to filter out shares from the "getAllShares" call which are no longer visible for the user. But I think we can solve it.

**Comment by rullzer** [comment]
After thinking about this I'm still not entirely sure what we want. I understand that it is weird that shares keep on exisiting but I also agree with @schiesbn that it is very weird when one delete cascades, especially since it is visible now.

For reshares assume the following situation:

```
Folder
|-subFolder
```

1. user0 shares Folder with user1
2. user1 shares Folder with user2
3. user2 shares subFolder with user3
4. user3 shares subFolder with user4

Now we have a number of views. We have the view from the owner of the files (user0). From people that share files and receive files. Thus:

user0:

```
Folder (shared by you with user1, shared by user1 with user2)
|-subFolder (shared by user2 with user3, shared by user3 with user4)
```

user1:

```
Folder (shared with you by user0, shared by you with user2)
|-subFolder
```

user2:

```
subFolder (shared with you by user1, shared by you with user3)
```


user3:

```
subFolder (shared with you by user2, shared by you with user4)
```


user4:

```
subFolder (shared with you by user4)
```

As you can see user0 has the entrire view. Since it is their data. The other users have a much more limited view. This is since they do not need to know everything (it is not their data). And it will become very complex and costly to calculate that all for larger installations.

No assume user0 thinks I don't want to share my data anymore with user1. Then they have to construct in their head the entrire share tree to know what consecuenses their removal has.

I agree with @schiesbn that probably only removing the share to user1. Would make the most sense. Since user0 has the complete overview anyway. And they can delete all the other shares as well. 

Deleting all the children was fine in the old situation since the owner of the files could only see their own shares. But now that they can see it all this makes it a bit more complex. I do think we should make sure that the behaviour is the same for reshares of `folder` and reshares of `subFolder`. So either we delete it all or we keep it all.

**Comment by icewind1991** [comment]
> Since user0 has the complete overview anyway

The problem is that with sharing sub-folders the re-share might be buried deep down into tree and easily missed

**Comment by DeepDiver1975** [comment]
What about disallowing to reshare sub folders of a share? @MTRichards @schiesbn @rullzer

**Comment by rullzer** [comment]
@DeepDiver1975 that would significantly simplify things. And in general I agree that complex resharing structures are more of a theoretical need.

But there is the following situation (which I think is pretty common):

* user0 shares a folder with user1
* user1 wants to make a subset (dir) available to someone.

The reason I know this is happening is because my file does stuff like this all the time. We have a shared folder of photos. And she wants to share some of those with other people (by link).

This would then no longer be possible.

**Comment by rullzer** [comment]
On the other hand we already do this limitation to some extend in the existing code.

```
fodler
|- subfolder (shared with me by someone)
```

If I receive the share of subfolder. And move it to `folder`. Then I can no londer share `folder` because it contains a share shared with me. So we do already limit this kind of stuff.

**Comment by MTRichards** [comment]
We can't remove child shared folder resharing, but can't we just add the reshare to the owners sub-folder as part of the share refactor? Just like if you reshare the entire folder? Then if you reshare a sub folder, it is just added to the owners list of sharees, 3 folders down, or whatever the situation is.

**Comment by rullzer** [comment]
@MTRichards that is the plan. (See the situation I described in https://github.com/owncloud/core/issues/19331#issuecomment-158873018).

However, that does not solve what to do on deletion.

**Comment by icewind1991** [comment]
I still feel deleting any reshares is the correct way

**Comment by schiesbn** [comment]
> The problem is that with sharing sub-folders the re-share might be buried deep down into tree and easily missed

Therefore you have the overview pages "shared with you", shared by you", etc.

The whole idea about getting rid of the re-share hierarchy was to make thinks easier. Otherwise we have the same mess again. What sounds relatively easy if we discuss this with three users (share + re-share) become quite complex if we take more levels of -re-shares into account and some user in between removes his share.

The whole idea of the flat sharing list was that you don't create a re-share in the sense of a tree but I give you the permission to create shares "in my name". Once you have done this there shouldn't be a difference weather I created the share by my own or you did it. The new share is a independent share. The share is there because I allowed other to create it, I as a owner can see and control it. That's it. 

I could even imagine that it would be quite confusing for the owner if he removes one share and 10 other shares disappear as magically. In the past this was different because he didn't saw the re-shares but now he sees them, so let him control it. I think in the most cases you want to remove exactly the share where you click on "unshare" if you see all shares and don't want to create a chain reaction deleting other shares as well. 

The good thing about the flat share list is that we no longer have to guess and try to make the right decision for the owner. The owner sees all shares and has the full control. So let him execute his power and don't get in the way. At least this would be my approach.

**Comment by PVince81** [comment]
> from my pov we will need string because of all discussed reasons - but this would require v2 of the sharing api which I'd like to postpone for now (maybe a topic for 9.1).

Agreed. Also changing int to string on the interfaces later shouldn't be a big change (or introducing new interfaces). The biggest part is making sure the clients can still work with string ids.

**Comment by DeepDiver1975** [comment]
Given the fact that a re-share is nothing more then a impersonated share of the owner - there is no need to handle the unshare operation any special.

A shares photos with B
B shares photos/vacation with C
For A C will be shown as shared with on photos/vacation
If A removes B from photos there is no need to remove C from photos/vacation automagically

**Comment by rullzer** [comment]
Added some text for the required db modifications (and why)

**Comment by jancborchardt** [comment]
@DeepDiver1975 yup. Actually if A removing B would automatically remove C as well that would be really confusing. After a few weeks/months no one remembers who reshared what with whom and unsharing from just one person should not cause that whole tree to collapse.

**Comment by jospoortvliet** [comment]
For what it's worth, I'd also vote with @DeepDiver1975 and @jancborchardt on this.

It is hard to figure out what people would expect and for sure some will have different needs and expectations anyway, but https://github.com/owncloud/core/issues/19331#issuecomment-158927029 by @schiesbn has the answer:

> The whole idea of the flat sharing list was that you don't create a re-share in the sense of a tree but I give you the permission to create shares "in my name". Once you have done this there shouldn't be a difference whether I created the share by my own or you did it. The new share is a independent share. The share is there because I allowed other to create it, I as a owner can see and control it. That's it.

It might not always be what users *want* but it does lead to the *most predictable* behavior. And the worst it will ever do is create a bit 'extra work' for the original sharer to manually remove all shares done by others he/she doesn't want.

**Comment by rullzer** [comment]
We must also make sure that the webinterface is ported to the OCS Share API in order to ensure a consistent view. As the old ajax endpoints still use the old code paths.

**Comment by moscicki** [comment]
Is this going to be part of OC9? Sharing currently is a major source of page loading slowness in current cernbox, we would like to check if this improves in OC9. 

CC: @karlitschek

**Comment by DeepDiver1975** [comment]
> Is this going to be part of OC9? Sharing currently is a major source of page loading slowness in current cernbox, we would like to check if this improves in OC9.

yes - definitely - parts have already been merged into master branch

**Comment by davivel** [comment]
@rullzer, I have a question 

What's the need to have a READ bit in permissions of a Share? Is there any situation where a valid Share should have a value 0 for that bit?

I am writing some automatic tests for updating permissions in the Android app and I thought that updating a Share to remove the READ permission would fail, but that's not the case.

Thanks.

**Comment by rullzer** [comment]
@davivel well it is there for legacy reasons. We used to allow removing it (which resulted in broken shares). So right now it is just always added. Even when you create a share without read permissions they will be added automatically.

If you try to remove them it will thus allow the call. But if you then fetch the share again the read permissions should still be there.

**Comment by davivel** [comment]
> Even when you create a share without read permissions they will be added automatically.

Nice to know :)

Thanks a lot, @rullzer

**Comment by rullzer** [comment]
@DeepDiver1975 I remember we talked about the bullet `allow to pass in a token when creating a new share` is that still required? I mean when update share allows you to migrate the share all should work right?

**Comment by PVince81** [comment]
@rullzer @schiesbn please move all remaining non-9.0 items to a separate ticket so we can close this one (if applicable)

**Comment by rullzer** [comment]
Ok lets move all remaining internal stuff to https://github.com/owncloud/core/issues/22209 and then this can be closed.
